resolwe/flow/managers/listener.py

```diff
@@ -73,9 +85,103 @@
 logger = logging.getLogger(__name__)
 
 
-class PythonProcessMixin:
+class ListenerPlugin(metaclass=abc.ABCMeta):
+    """Base class for listener plugin classes."""
+
+    @classmethod
+    def __init_subclass__(cls: Type["ListenerPlugin"], **kwargs):
+        """Register class with the registry on initialization."""
+        super().__init_subclass__(**kwargs)
+        ListenerPlugins().add_plugin(cls)
+
+    def _get_handler_name(self, command_name: str) -> str:
+        """Get the handler name."""
+        return f"handle_{command_name}"
+
+    def get_handler(
+        self, command_name: str
+    ) -> Optional[Callable[[Message, "Processor"], Response]]:
+        """Return the plugin handler
+
+        When so such plugin exists return None.
+        """
+        return getattr(self, self._get_handler_name(command_name), None)
+
+
+class ListenerPlugins(MutableMapping[str, object]):
+    """Class for registering listener plugins."""
+
+    __instance = None  #  A single instance of this class
+    _plugin_classes: Dict[str, Type[ListenerPlugin]] = dict()
+    _plugins: Dict[str, ListenerPlugin] = dict()
+
+    @classmethod
+    def __new__(cls, *args, **kwargs):
+        """Create or return singleton."""
+        if ListenerPlugins.__instance is None:
+            ListenerPlugins.__instance = super().__new__(cls)
+
+        return ListenerPlugins.__instance
+
+    def _get_class_name(self, plugin_class: Type[ListenerPlugin]) -> str:
+        """Get the full class name."""
+        return "{}.{}".format(plugin_class.__module__, plugin_class.__name__)
+
+    def add_plugin(self, plugin_class: Type[ListenerPlugin]):
+        """Add new plugin class.
+
+        :raises RuntimeError: when plugin is already registered.
+        """
+        plugin_name = self._get_class_name(plugin_class)
+        assert (
+            plugin_name not in self._plugin_classes
+        ), "Plugin named {plugin_name} already registered."
+        self._plugin_classes[plugin_name] = plugin_class
+
+    def get_handler(self, command_name: str) -> Optional[Callable[[Message], Response]]:
+        """Return the handler for the given command.
+
+        When such handler exists in multiple plugins one is chosen at random.
```

Comment:
This may be a not so uncommon use-case as plugins will be defined in different packages. It would be great to iterate over all plugins and raise an error if two handlers are found.

Reply:
OK.